crates/holochain_zome_types/src/countersigning.rs

```diff
@@ -0,0 +1,288 @@
+pub struct CounterSigningAgentState {
+    /// The index of the agent in the preflight request agent vector.
+    agent_index: u8,
+    /// The current (frozen) top of the agent's local chain.
+    chain_top: HeaderHash,
+    /// The header sequence of the agent's chain top.
+    header_seq: u32,
+    /// The signature of all preflight request data and agent state data.
+    signature: Signature,
+}
+
+impl CounterSigningAgentState {
+    /// Agent index accessor.
+    pub fn agent_index(&self) -> u8 {
+        self.agent_index
+    }
+
+    /// Chain top accessor.
+    pub fn chain_top_ref(&self) -> &HeaderHash {
+        &self.chain_top
+    }
+
+    /// Header seq accessor.
+    pub fn header_seq(&self) -> u32 {
+        self.header_seq
+    }
+
+    /// Signature accessor.
+    pub fn signature_ref(&self) -> &Signature {
+        &self.signature
+    }
+}
+
+/// A vector of agents to countersign a shared entry.
+/// The vector must be sorted to generate a CounterSigningTag.
+#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
+pub struct CounterSigningAgentStates(Vec<CounterSigningAgentState>);
+
+/// Enum to mirror Header for all the shared data required to build session headers.
+/// Does NOT hold any agent specific information.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub enum HeaderBase {
+    /// Mirrors Header::Create.
+    Create(CreateBase),
+    /// Mirrors Header::Update.
+    Update(UpdateBase),
+    // @todo - These headers don't have entries so there's nowhere obvious to put the CounterSigningSessionData.
+    // Delete(DeleteBase),
+    // DeleteLink(DeleteLinkBase),
+    // CreateLink(CreateLinkBase),
+}
+
+/// Base data for Create headers.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CreateBase {
+    entry_type: EntryType,
+    entry_hash: EntryHash,
+}
+
+/// Base data for Update headers.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct UpdateBase {
+    original_header_address: HeaderHash,
+    original_entry_address: EntryHash,
+    entry_type: EntryType,
+    entry_hash: EntryHash,
+}
+
+/// All the data required for a countersigning session.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CounterSigningSessionData {
+    preflight_request: PreflightRequest,
+    responses: Vec<CounterSigningAgentState>,
+}
+
+/// Build an unsigned Create header from session data, shared create data and an agent's state.
+impl
+    From<(
+        &CounterSigningSessionData,
+        &CreateBase,
+        &CounterSigningAgentState,
+    )> for Create
+{
+    fn from(
+        (session_data, create_base, agent_state): (
+            &CounterSigningSessionData,
+            &CreateBase,
+            &CounterSigningAgentState,
+        ),
+    ) -> Self {
+        Create {
+            author: session_data.preflight_request.signing_agents[agent_state.agent_index as usize]
```

Comment:
This is fallible because agent_index is user-defined, right? We don't want to crash on bad input. So should this function should return Result or Option?